### PR TITLE
add decimal typehints and converter

### PR DESCRIPTION
**Pull request by HKandulla**

### CODE REVIEW

File: src/Phpro/SoapClient/Soap/TypeConverter/DecimalTypeConverter.php [hunk 1]
```diff
@@ -0,0 +1,54 @@
...
+    /**
+     * {@inheritdoc}
+     */
+    public function convertXmlToPhp($data)
+    {
+        $doc = new DOMDocument();
+        $doc->loadXML($data);
+
+        if ('' === $doc->textContent) {
+            return null;
```
Comment (veewee) [inline src/Phpro/SoapClient/Soap/TypeConverter/DecimalTypeConverter.php]: Should this method really return null? I would think that if it is null, it would return the value `0.0`. Otherwise we won't be able to add strict types to the generated code in the future. Or do you see use-cases for `null` values?

File: test/PhproTest/SoapClient/Unit/TypeConverter/DecimalTypeConverterTest.php [hunk 2]
```diff
@@ -0,0 +1,81 @@
+<?php
+
+namespace PhproTest\SoapClient\Unit\Middleware;
```
Comment (veewee) [inline test/PhproTest/SoapClient/Unit/TypeConverter/DecimalTypeConverterTest.php]: It looks like you've placed this test in a wrong namespace.

File: test/PhproTest/SoapClient/Unit/TypeConverter/DecimalTypeConverterTest.php [hunk 3]
```diff
@@ -0,0 +1,81 @@
...
+     */
+    public function testConvertPhpToXml()
+    {
+        $xml = '<decimal>24.7</decimal>';
+
+        $output = $this->converter->convertPhpToXml((float) 24.700);
+
+        $this->assertSame($xml, $output);
+    }
+}
```
Comment (veewee) [inline test/PhproTest/SoapClient/Unit/TypeConverter/DecimalTypeConverterTest.php]: Can you add a newline at the EOF?

**Comment by veewee** [comment]
Hi @HKandulla,

Thanks for the PR! I've added some small remarks.
You can also add a double converter here.

File: src/Phpro/SoapClient/Soap/TypeConverter/DecimalTypeConverter.php (hunk 1 above)
Comment (HKandulla) [inline src/Phpro/SoapClient/Soap/TypeConverter/DecimalTypeConverter.php]: I was thinking about this too. I think null should be possible in all converter (no value) and strict types should always be nullable (?float), unfortunately this would make PHP 7.1 a requirement. In my case the service uses null to reset the value to its default. What do you think?

Comment (veewee) [inline src/Phpro/SoapClient/Soap/TypeConverter/DecimalTypeConverter.php]: What do you mean with "reset the value to its default"?

**Comment by HKandulla** [comment]
HI @veewee,
thanks for your remarks. I make the changes and also add a double converter.

File: src/Phpro/SoapClient/Soap/TypeConverter/DecimalTypeConverter.php (hunk 1 above)
Comment (HKandulla) [inline src/Phpro/SoapClient/Soap/TypeConverter/DecimalTypeConverter.php]: The SOAP-Service I am working with resets a property to its default value if I the corresponding tag is empty. If the tag has the value 0.0 it would set this value (and not the default)..

Comment (veewee) [inline src/Phpro/SoapClient/Soap/TypeConverter/DecimalTypeConverter.php]: For me the value `null` is also OK, but we should take this in consideration if we want to support strict types at some point. Let's keep it this way for now :)

Comment (HKandulla) [inline src/Phpro/SoapClient/Soap/TypeConverter/DecimalTypeConverter.php]: great. Thanks.

File: src/Phpro/SoapClient/ClientBuilder.php [hunk 4]
```diff
@@ -88,6 +88,7 @@ public function __construct(ClientFactoryInterface $clientFactory, $wsdl, array
         // Add default converters:
         $this->addTypeConverter(new TypeConverter\DateTimeTypeConverter());
         $this->addTypeConverter(new TypeConverter\DateTypeConverter());
+        $this->addTypeConverter(new TypeConverter\DecimalTypeConverter());
```
Comment (veewee) [inline src/Phpro/SoapClient/ClientBuilder.php]: I guess we should also add the double type converter to the defaults?

Comment (HKandulla) [inline src/Phpro/SoapClient/ClientBuilder.php]: oh yes, sorry. It's added.

**Comment by veewee** [comment]
Thanks for the work already. It looks like there is one thing left to do: updating the documentation.
https://github.com/phpro/soap-client/blob/master/docs/type-converter.md#convert-soap-types

**Comment by HKandulla** [comment]
It's a pleasure. It's a great library! Is there anything else I can do?

**Comment by veewee** [comment]
@HKandulla, Sure you can. There is the issue you created about ignoring null values (https://github.com/phpro/soap-client/issues/34)
Besides that, you can pick any open issue with the label [help wanted](https://github.com/phpro/soap-client/issues?q=is%3Aissue+is%3Aopen+label%3A%22help+wanted%22) you want or add something you are still missing to the package.

**Comment by HKandulla** [comment]
I am still busy with the tests on #34. I want to add proper tests that make sure that even nodes with empty child nodes are removed by my middleware. But I definetely do this this week.. I was thinking about adding support for sequences as in:

`<s:complexType name="ArrayOfSyncCustomerRecord">
  <s:sequence>
<s:element minOccurs="0" maxOccurs="unbounded" name="SyncCustomerRecord" nillable="true" type="tns:SyncCustomerRecord"/>
</s:sequence>
</s:complexType>
`
But I need to investigate a little more. Any maybe the IteratorInterface is a good enough solution. If you have am idea of what is needed as far as support for sequences/arrays goes, please let me know.

**Comment by veewee** [comment]
For sequences, I always use the Iterator. It works well and you can just iterate over the results as you probably always want to do. Let's continue the discussion in #34.
